ENH: Create helper for conversion to arrays

seberg opened the pull request:
Had to give this a start to be able to experiment.  It does solve at least some problems (e.g. `np.select()` use-case and it looks like it would fix and/or simplify linalg code).

The main thing is that we often convert multiple arrays, and then want something like:
* `a, b, c = np.asarray(a), np.asarray(b), np.asarray(c)` and then `dtype = np.result_type(a, b, c)` often with an additional `0.0`.  That is a pretty common pattern, NEP 50 makes it tricky if you wish to support weak scalars, because you need to call `result_type` with the scalar information intact (before the `asarray` or manually undo).  (For example, this happens in `np.select()`)
  * Maybe libraries can disregard that, but for NumPy it's hard, since you want many functions to be ufunc-like, so if ufuncs do it, others probably should be able to/try to.
* Maybe mainly internally we sometimes need the `__array_wrap__` for subclass support, although it seems consistently used for single input functions (outside ufuncs).
  * Additional motif:  A clean wrapping pattern seems required for making scalar logic sane.
* It  may be useful to know if an input was a scalar or not (not a 0-D array).  You need to ask NumPy about it, but it discovered at conversion time.  So it is an information that is weird to tag onto an array (probably?) but figuring it out later would require duplicating work.

First, I wanted to just make a single function.  But with multiple things above that may be interesting to return, and e.g. `result_type()` might error and I am not sure we always need it.

So, is it insane to create a rich object (rather than a complicated return value and many kwargs):
```
# better names of course, currently in `np.core._multiarray_umath.ArrayCreationHelper`
arrays = creation_helper(*inputs)

# e.g. a linspace could do something like this:
dtype = arrays.common_dtype(ensure_inexact=True)

arrays.arrays  # arrays as tuple (probably allow arrays[0])

# Apply the typical wrapping if needed
# (should cache wrap, or fetch it early on rather, maybe just
# make it a method which applies the wrapping `arrays.wrap(result)`).
if arrays.wrap is not None:
    return arrays.wrap(result)
return result 
```
(I can see `a, b, info = helper(a, b)` as a nice pattern; OTOH it makes including casting kwarg heavy since you somewhat need to include it in the first step.)

Further things might include:
* Casting to a given or the common dtype.  (maybe `arrays.to_common_dtype(ensure_inexact=True)`)
* I think @mdhaber was hoping to include broadcasting.  Not sure, since it is one call away, but also seems fine to add.

mdhaber commented:
1.
> I think @mdhaber was hoping to include broadcasting. Not sure, since it is one call away, but also seems fine to add.

That would be nice. It seems like a common enough follow-up step that it would be good to add to this convenience function.

2. > So, is it insane to create a rich object...

We do this frequently in SciPy now, so not insane. Seems unusual for NumPy, but I wouldn't complain. 

3. Idea: maybe instead of an extra parameter `ensure_inexact=True`, could `extra_dtype` accept a class like `np.inexact`? 

4. I'm not used to reading C. Would it be easy to add a docstring or the equivalent sequence of existing NumPy commands so I can check my understanding of what is going on?

Thanks for this!

seberg commented:
Yeah, agree that the ufunc machinery should share as much as possible, although was a bit hoping to mainly consolidate Python paths first.  The reason was that looking at the array-wrap implementation in ufuncs they are messy and buggy (there is an issue), so fixing that seems like a bit of a messy change easier on its own.

I think a method for output array prep could make sense, adding it to the initial call seems like it would overload meaning (we don't know the common dtype is right for one thing).

> Idea: maybe instead of an extra parameter ensure_inexact=True, could extra_dtype accept classes like np.inexact, if that makes sense?

Right, that is how it is implemented.  Just `dtype=np.inexact` doesn't quite work yet and I thought the `result_type(..., 0.)` pattern is so common, it might actually make sense to have it explicitly.

> 5. [discussed] adding axis to be ignored when broadcasting for use in, say, multi-sample reduction statistics [...]

Ah, right!  Yeah, maybe make its own issue, it seems to me we should with integrating that into `np.broadcast_arrays` first.  Admittedly, I think that may require reimplementing the whole function (maybe for better).

rgommers commented:
@seberg from your description it is not entirely clear to me if you mean for this to be a public or private function? With the namespace we discussed somewhere else, `np.lib.array_utils` I believe, we should get a place to put utilities like this - but I can't tell from the description yet if this would be useful more widely.

Since it was mentioned above, we recently came across the need for `normalize_axis_index` in SciPy - that's also a useful utility, which is not public right now.

seberg commented:
The first issue I want to solve would mean private is fine, but yes suspect we should just make it public and if we are unsure about some methods/attributes or so give it an `_` (which would mean putting it into `np.lib.array_utils` namespace, or different name if anyone suggests a better one soon).
(And @mhvk's interest makes me lean more towards public at last if I can get some of what @mdhaber wants too.)

mdhaber commented:
I just ran into this little gem that might be avoided with use of a helper like this.
```python3
np.float16([2]) < np.float32(2.0009766)  # array([False])
np.float16([2]) < np.float32([2.0009766])  # array([ True])
```

seberg commented:
You might think that helps, but it probably wouldn't :).  This is NEP 50 stuff.

mdhaber commented:
If before comparison the float16 is promoted to the result type or if the float32 is broadcasted to the result shape, the comparison is done correctly.

seberg commented:
@mdhaber no, `np.result_type()` would do the same *down* promotion in the above case until you do NEP 50 .

mdhaber commented:
I always pass the dtypes into `result_type` instead of the arrays.
```
import numpy as np
x = np.float16([2])
y = np.float32(2.0009766)
print(x < y)
x1, y1 = np.broadcast_arrays(x , y)
print(x1 < y1)
dtype = np.result_type(x.dtype, y.dtype)
print(x.astype(dtype) < y.astype(y))
```
```
[False]
[ True]
[ True]
```

seberg commented:
Right, if you pass the dtypes you get the right thing, but that is the same as adopting NEP 50.

seberg commented:
I was trying to get back to this a bit, but having difficulties to quite formulate what we need/want here.  I would prefer to not go the feature-creep route for starters.

Things that I would like:
* Allow converting multiple arrays at once more conveniently (easy).
* The result should have `common_dtype()` as it is pretty useful especially with NEP 50.
* The result should have a `wrap()` to apply `__array_wrap__()` when it makes sense. [^2]
* We could have more options:
  * `shape` (braodcast shape)
  * A `broadcast()`  method to modify the "arrays"` (probably needs a local implementation for starters, which then might be adoptable elsewhere).  Broadcasting _could_ be a kwarg also.

There are some options, which would probably fit as kwargs [^1] :
* `subok=True` (to set it to `False`)
* broadcasting *could* be a kwarg.

Another API, I worry about more: it would be nice to provide a clean way to preserve Python better.  This is to allow later normal promotion to use NEP 50 rules.
Due to the `isclose` and similar use-case, it also seems to me that in many cases it would make sense for broadcasting to *not* modify scalars (or 0-D objects).
For that, it would make sense to either have a kwarg `preserve_scalars=True` or an attribute `arrays_or_scalars`?

Maybe one solution is to make it:
```python
conv = array_converter(*arrays)
dt = conv.common_dtype()  # always applicable, may do initial conversion to arrays
conv.shape  # could do this to get the broadcast shape, but could also defer.
conv.was_pyscalar  # maybe.  Or an enum with `0-D, scalar, python scalar`.

arrays = conv.arrays(subok=True, allow_scalars=True, broadcast=True, dtype=dt)

# Wrap result (if wanted) (see footnote 2) [^2]
return conv.wrap(result)
```
Any thoughts?  Implementing any of this isn't super hard, but getting more clarity on where to start without a feature creap, is for me.

---

I think, I like that last thought as API (initialization takes no kwargs, but fetching the arrays is a function and not an attribute/property).  But it isn't quite clear to me.  Especially how to do scalar preservation.

[^1]: kwargs are slow for classes due to `tp_new` pre Python 3.13, but lets not worry about it
[^2]: We currently mainly/only seem to use wrap if `subok=False`, so the details here are a bit tricky (should you ever call it when you already got a subclass?).
However, one reason I would like a clear entry-point is that it would help *a lot* thinking about gh-24897, since it could take a kwarg like `preserve_scalars=True`.

seberg commented on numpy/_core/function_base.py:
```diff
@@ -126,21 +127,20 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
-    stop = asanyarray(stop) * 1.0
+    conv = _array_converter(start, stop)
+    start, stop = conv.as_arrays(subok=True)
+    dt = conv.result_type(ensure_inexact=True)
```
Iterated a bit, this is still a bit rough (I would like to consolidate `__array_wrap__` a bit at least).

This does define `conv.wrap()` as the one true way to call `__array_wrap__`, but for NEP 50, the interesting part is really this part.
(I haved the cast to float into the subtract via `dtype=`, which isn't strictly necessary.)

The point is, that this means `np.linspace(np.float32(3), 10)` returns float32.  (yes, still needs tests, but thought I would note this, as this pattern would hopefully be useful a few places, that are currently subtly broken in the same way.)

seberg replied:
@mhvk beyond cleanup, part of my thought for consolidating is that it should help with gh-24897.  Because wrapping is a reasonable place to decide scalar vs. array logic.

mhvk commented on numpy/lib/_shape_base_impl.py:
```diff
@@ -403,17 +406,8 @@ def apply_along_axis(func1d, axis, arr, *args, **kwargs):
-        # matrices have to be transposed first, because they collapse dimensions!
-        out_arr = transpose(buff, buff_permute)
-        return res.__array_wrap__(out_arr)
+    res = transpose(buff, buff_permute)
```
For subclasses other than `matrix`, this would be an API change if `.transpose()` is overridden. That seems rather unlikely (and wrong-headed), so I'm fine with making the change regardless...

mhvk reviewed (commented):
@seberg - now had a better look. I think this is really nice and it should be both relatively straightforward to extend and to use some of it inside the ufuncs.

One suggestion: it might be nice if `conv` can be indexed like a tuple. That should then return the "anyarray" version (in which case the default `subok=False` for `as_arrays()` makes sense; without this, I feel `subok=True` would be a more logical default).

If not doing that, I would suggest to replace `ary, = conv.as_arrays()` with `ary = conv.as_arrays()[0]` for clarity (the trailing comma before the equal sign is easy to miss...).

seberg reviewed (commented):
I have addressed all the comments and changed the default for `subok` (I was torn on that initially, but I made my peace with it).

I allow indexing now (which also allows `a, b = conv` unpacking, although calling the method is actually a tiny bit faster (the reason is that unpacking checks `conv[length]` to trigger the index error).

---

I have a start on reorganizing the ufuncs to share some infrastructure for array-wrap at least.  But decided that shouldn't stop us here, since it mostly means moving the `find_wrap` function out to share it and the ufunc changes are not quite utterly trivial.

mhvk commented on numpy/_core/function_base.py:
```diff
@@ -126,21 +127,20 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
 
-    delta = stop - start
+    # Use `dtype=type(dt)` to enforce a floating point evaluation:
+    delta = np.subtract(stop, start, dtype=type(dt))
```
I thought the replacement of `start = start * 1.0` was `.result_type(ensure_inexact=True)`? If it isn't necessary, I would just keep `delta = stop - start`.

mhvk replied:
OK, that makes sense, resolving.

mhvk commented on numpy/_core/function_base.py:
```diff
@@ -180,10 +180,11 @@ def linspace(start, stop, num=50, endpoint=True, retstep=False, dtype=None,
         _nx.floor(y, out=y)
 
+    y = conv.wrap(y.astype(dtype, copy=False))
     if retstep:
```
Possibly worth simplifying to `return (y, step) if retstep else y`

mhvk commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,558 @@
+        }
+        /* At this point, assume cleanup should happen for this item */
+        self->narrs++;
+        Py_INCREF(item->object);
```
Shouldn't the INCREF be before `narrs++`?

mhvk replied:
Ah, I see: the fast path were things are set and INCREF'd cannot fail, so this will always work. Looks good!

mhvk commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,558 @@
+        }
+    }
+
+    Py_INCREF(self);
```
Are you sure you need the `INCREF` here? `PyObject_NewVar` returns a new reference already (which you `DECREF` in the fail path).

mhvk commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,558 @@
+        return NULL;
+    }
+
+    /* When we only have scalars, the default is to retur an array: */
```
`retur` -> `return`

mhvk commented on numpy/lib/_function_base_impl.py:
```diff
@@ -5178,14 +5179,9 @@ def delete(arr, obj, axis=None):
-        except AttributeError:
-            pass
+    conv = _array_converter(arr)
+    arr, = conv.as_arrays(subok=False)
```
Why this had `asarray` rather than `asanyarray` -- o well, here definitely ensure no change in API!

Though perhaps still do `arr = conv.as_arrays(subok=False)[0]`?

mhvk replied:
OK, this is clearly in the eye of the beholder!

mhvk reviewed (commented):
Looks mostly great now! Just a few remaining comments.

One comment: with the indexing giving any array, I think it is actually OK to have `subok=False` as default for `as_arrays()`. Though also completely happy to keep it as you have it now - in principle, subclasses should behave like regular ones, so it should be fine to keep them (and it is nice if the default actually corresponds to the least amount of work, which is the case for `subok=True` since you are not viewing as a regular array).

mhvk commented:
p.s. On the `asanyarray`, beyond the rationale that subclasses should behave the same way and thus not be converted (of course, they don't always do...), it is certainly nice if generally they get passed on, since then `__array_ufunc__` coverage often is good enough for functions to work. For `Quantity` at least, there are still a lot of functions where we do not have to override; see https://github.com/astropy/astropy/blob/main/astropy/units/quantity_helper/function_helpers.py (the `SUBCLASS_SAFE_FUNCTIONS` set; as you'll notice, there are also a lot that do need overrides, which means a very long list of rather boilerplate code that is little fun to write, so all the better if some functions don't need it!)

seberg reviewed (commented):
Yeah, about `SUBCLASS_SAFE_FUNCTIONS`:  it would be great if we could formalize this in some form and document the pattern.  But the best would be to generalize it to non-subclasses somehow, which might be just calling `func._implementation()` rather than `ndarray.__array_function__`.

> with the indexing giving any array, I think it is actually OK to have subok=False as default for

I agree more with the latter part, unless we had `asanyarray` also: align default and short-hand.  Plus, if you want a real nudge toward `subok` having the opposite default is no good.

The main reason for *not* doing it was maybe always matrices (and probably historic).  But we should hopefully worry less and less about those.

Also added a start for documentation.

ngoldbaum commented on numpy/_core/src/multiarray/ctors.c:
```diff
@@ -1541,12 +1543,19 @@ PyArray_FromAny(PyObject *op, PyArray_Descr *newtype, int min_depth,
+ * This means it was:
+ * - Recognized as a scalar by a/the dtype.  This can be DType specific,
+ *   for example a tuple may be a scalar, but only for structured dtypes.
+ * - Anything not recognized a scalar (mapped to a DType) but also not
```
```suggestion
 * - Anything not recognized as an instance of a DType's scalar type but also not
```

I got tripped up on the original wording - take it or leave it, but if you leave it, it should be `recognized as a scalar`

ngoldbaum commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,563 @@
+            }
+            double curr_priority = PyArray_GetPriority(obj, 0);
+            if (wrap == NULL || priority < curr_priority) {
+                Py_XSETREF(wrap, new_wrap);
```
You can't get here again after the first time you enter this block. Maybe `break` at the end to enforce that we don't need to do further loop iterations?

mhvk replied:
I think you can get here if the priority of the next input is higher than what was there before.

ngoldbaum replied:
Argh, misread, you're right.

ngoldbaum commented on numpy/_core/_add_newdocs.py:
```diff
@@ -4713,6 +4713,86 @@
+    Helper to convert one or more objects to arrays.  Integrates machinery
+    to deal with the ``result_type`` and ``__array_wrap__``.
+
+    The reason for this is that e.g. ``result_type`` needs to covert to arrays
```
```suggestion
    The reason for this is that e.g. ``result_type`` needs to convert to arrays
```

ngoldbaum reviewed (commented):
I looked this over and spotted a few minor things but no showstoppers.

mhvk commented on numpy/_core/_add_newdocs.py:
```diff
@@ -4713,6 +4713,86 @@
+add_newdoc('numpy._core._multiarray_umath', '_array_converter', ('scalar_input',
+    """
+    A tuple which is ``True`` if the original inputs were scalars and
+    ``False`` otherwise.  Scalars are all objects which are coerced to a
```
How about
```
A tuple which indicates for each input whether it was a scalar that
was coerced to a 0-D array (and was not already an array or something
converted via a protocol like ``__array__()``).
```

mhvk commented on numpy/_core/_add_newdocs.py:
```diff
@@ -4713,6 +4713,86 @@
+    ``False`` otherwise.  Scalars are all objects which are coerced to a
+    0-D array but are not already arrays or converted via one of the protocols.
+
+    .. note::
```
My sense would be to delete this note: it probably adds more confusion than it solves...

mhvk commented on numpy/_core/_add_newdocs.py:
```diff
@@ -4713,6 +4713,86 @@
+    ----------
+    arr : ndarray
+        The object to be wrapped.  Ideally, this is always an ndarray
+        or subclass, although NumPy scalars are accepted as of now by
```
How about
```
The object to be wrapped. Normally an ndarray or subclass,
although for backward compatibility NumPy scalars are also accepted
(these will be converted to a NumPy array before being passed on to
the ``__array_wrap__`` method).
```

mhvk commented on numpy/_core/_add_newdocs.py:
```diff
@@ -4713,6 +4713,86 @@
+        When ``True`` will convert a 0-d array to a scalar via ``result[()]``
+        (with a fast-path for non-subclasses).  If ``False`` the result should
+        be an array-like (as ``__array_wrap__`` is free to return a non-array).
+        By default (``None``), a scalar is returned if all inputs are scalar.
```
are scalar -> were scalar

mhvk commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,563 @@
+        PyObject *obj = self->items[i].object;
+        if (PyArray_CheckExact(obj)) {
+            if (wrap == NULL || priority < 0) {
+                Py_INCREF(Py_None);
```
Strictly, you don't need to set the wrap here, given the `if (wrap == NULL)` clause at the end. (same for scalar)

mhvk commented on numpy/lib/_arraysetops_impl.py:
```diff
@@ -113,14 +116,15 @@ def ediff1d(ary, to_end=None, to_begin=None):
     np.subtract(ary[1:], ary[:-1], result[l_begin:l_begin + l_diff])
-    return result
+
+    return conv.wrap(result)
```
Ah, yes, this way we don't require that any non-subclass overrides `empty_like` via `__array_function__` -- makes sense, so resolving.

mhvk reviewed (commented):
I think what's left is only typos and some suggestions for the new docstrings!

mhvk commented:
p.s. Probably time for some squashing of commits too...

seberg commented:
I don't want to jinx this because of it, but the `[()]` logic for subclasses doesn't work for current `memmap` (which incorrectly always does it).
Now that is maybe even never used right now, but it makes me wonder if we need another approach, which might mean passing `scalar=True` in `__array_wrap__`!?

ngoldbaum commented:
see https://github.com/numpy/numpy/issues/25278 which was reported with impressive timing for this PR

mhvk commented:
I'm still not quite sold on `return_scalar`, but I'm also realizing I've lost what little overview I had of the problems that are caused by the various steps. This makes me wonder if it would be better to have a PR first that creates a helper that simply keeps the status quo, i.e., makes no attempt to be clever on wrapping, just reproducing the `__array_wrap__` calls that are present now.

seberg committed:
MAINT: Allow PyArray_FromAny_int to signal if the input was a scalar

seberg committed:
ENH: Add private array_conversion helper

This helper allows dealing with result_type and array wrapping
without explicitly holding on to the original objects and
especially potentially converting them many times.

seberg committed:
MAINT: Use new converter to fix promotions and avoid manual wrapping

This doesn't quite fix all places, but it does most (most prominently
linalg also does some quite awkward wrapping).

seberg commented:
@mhvk and @ngoldbaum, I merged main (i.e. the array-wrap helper) and applied all the remaining doc suggestions (thanks, they were great!).  Then squashed it down to three commits.

I _think_ this should be pretty much ready now, since IIRC it was relatively settled.  But happy to do another round if things come up of course.

mhvk commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,471 @@
+/*
+ * This file defines an _array_converter object used internally to NumPy
```
"used internally *in* Numpy *to* [deal ...]"

mhvk commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,471 @@
+        item->DType = NPY_DTYPE(PyArray_DESCR(item->array));
+        Py_INCREF(item->DType);
+
+        if (npy_mark_tmp_array_if_pyscalar(
```
Could this be `if (item->scalar_input && npy_mark...)`? If so, maybe combine with l.93-95,
```
if (item->scalar_input) {
    if (npy_mark...) {
        ...
    }
    else {
        ...
    }
}
else {  /* non-scalar input */
    self->flags &= ~(NPY_CH_ALL_PYSCALARS | NPY_CH_ALL_SCALARS);
}
```

seberg replied:
Yes, that is correct, and I adopted it, but not as suggested here.  The pretty-code annoyance is that `item->descr` needs to be dealt with, so whatever you do, you have some duplication.

But I agree, it isn't super nice to split this from L93, and maybe easier to grok if it is clear that `npy_mark_tmp_array_if_pyscalar` does indeed require `scalar_input`.

mhvk commented on numpy/_core/_add_newdocs.py:
```diff
@@ -4728,6 +4728,81 @@
+    float, or complex.
+    """)
+
+add_newdoc('numpy._core._multiarray_umath', '_array_converter', ('scalar_input',
```
I think `scalar_input` is not used anywhere in the code. Do you have cases in mind where it would be useful? It sort-of feels like it should be, but if there is no use, should it be exposed? (Fine to leave in, just asking.)

Also, right now it is a property. Might it make sense to make it a method (with no arguments), so that one could extend it if needed? (E.g., to ask not just for scalar, but python scalar inputs.)

But perhaps all this just argues for keeping `_array_converter` private for now, so we can see what is best!

ngoldbaum replied:
The linter says this line is one character too long

seberg replied:
I didn't think we cared, but fair, pushed the fix (inclued it in the previous commit).

mhvk commented on numpy/_core/src/multiarray/array_converter.c:
```diff
@@ -0,0 +1,471 @@
+        return NULL;
+    }
+
+    /* When we only have scalars, the default is to return an array: */
```
I was again a bit confused about this, maybe it helps to say "We follow the default setting for as_arrays of CONVERT_IF_NO_ARRAY"

mhvk commented on numpy/_core/_add_newdocs.py:
```diff
@@ -4728,6 +4728,81 @@
+    ----------
+    subok : True or False, optional
+        Whether array subclasses are preserved.
+    pyscalars : {"convert", "preserve", "convert_if_no_array"}, optional
```
This argument is not used anywhere, so a bit of the same question as for `scalar_input` above. In particular, I wonder if it wouldn't be handier to pass this on *input* (perhaps still allowing to override it here). But without a use case, probably not worth overthinking it now!

mhvk reviewed (commented):
This looks good! Looking at it again, I mainly wonder about methods and arguments that are defined but not actually used. Arguably, we should not have those without a clear use case... But as long as this is private, perhaps it is OK?

seberg committed:
MAINT: (array_converter) Address smaller review comments by Marten

seberg commented:
Addressed the smaller comments, thanks.

The API issues are not very clear to me.  I really like this type of feature, but it is quite fuzzy to me how it should look like and which details are useful (enough).

That is why I opted for private for now:  I am not comfortable with saying that this API seems good (enough) unless others give very clear +1's.

Generally, I would like this or something similar to be public, because I think it can be generally useful and also specifically because I think it is more useful with NEP 50 if downstream wants to be pedantic.

From a maintenance burden I am not super worried even if we deprecate it again soonish, we would name squat `np.lib.array_utils.array_converter`, though.

---

If we **keep things private**, I somewhat like just keeping those odd features.  I can see the `scalar_input` to be useful at least in niches like a python searchsorted like:
```
conv = array_converter(haystack, needle)
haystack, needle = conv

# The needle decides whether a scalar result makes sense:
return conv.wrap(result, to_scalar=conv.scalar_input[1])
```
I am not sure that we will use it (in fact I doubt it in the near future), but I also like the idea of "advertising" that this is possible.
(Part of why it is there, was that it is very easy and I also thought it may be useful for the python scalar preservation, which the conversion already does now and it seems to work well.)

If we lean towards **public API**, I fully agree to ditch them all until it is very clear we or downstream use them.  Those odd features shouldn't stop us from making this public!

mhvk reviewed (approved):
OK, code looks good now! On the API, I'm fine with keeping it private for now and seeing what ends up being useful and what not. I'll approve now, but won't merge to give @ngoldbaum a chance to look.

p.s. Note that the linter failed.

ngoldbaum commented:
I'll look at this tomorrow.

ngoldbaum reviewed (commented):
I looked over the C code again and didn't spot any issues. Please push a fixup for the linter then this should be good to go.

ngoldbaum commented:
I figure it's better to keep the linter happy so people don't get surprise linter corrections on future PRs. Thanks!
